### PR TITLE
Reduce peak gpu memory usage and support moe estimation

### PR DESCRIPTION
**Pull request by xin3he**
- Reduce peak memory usage by calling clear_memory cosidering performance effort. 
- Move best_params to CPU and make sure clear memory before moving back.
- [move loss device to the second card if card_0_in_high_risk](https://github.com/intel/auto-round/pull/981/commits/a27e1dd042d6e19a930a09da6f07ddd0b4c17ce1)
- support Deepseek R1 W4A16 tuning with 3 CUDA cards (80GB) （--enable_torch_compile)
- support llama3.3 70B W4A16 tuning with 2 Intel GPU cards (24GB)（--enable_torch_compile)

### CODE REVIEW

File: auto_round/compressors/utils.py [hunk 1]
```diff
@@ -199,13 +199,16 @@ def check_awq_gemm_compatibility(model, bits, group_size, sym, layer_configs=Non
...
-def collect_best_params(block):
+def collect_best_params(block, low_gpu_mem_usage: bool = False):
     params = {}
     for n, m in block.named_modules():
         if hasattr(m, "orig_layer"):
             params[n] = {}
             for key in m.params.keys():
-                params[n][key] = copy.deepcopy(m.params[key].data)
+                if low_gpu_mem_usage:
+                    params[n][key] = m.params[key].data.cpu()
```
Comment (wenhuach21) [inline auto_round/compressors/utils.py]: if the data is already on cpu like cpu tuning, this will causes some issue

File: auto_round/compressors/base.py [hunk 2]
```diff
@@ -2760,6 +2770,7 @@ def _quantize_blocks(
                 modules = [get_module(model, n) for n in names]
                 m = WrapperMultiblock(modules)
 
+            m.config = model.config if hasattr(model, "config") else None
```
Comment (wenhuach21) [inline auto_round/compressors/base.py:2801]: del this after tuning

File: auto_round/compressors/utils.py (hunk 1 above)
Comment (xin3he) [reply auto_round/compressors/utils.py]: cpu() will copy and it's expected behavior

File: auto_round/utils/device.py [hunk 3]
```diff
@@ -732,6 +731,75 @@ def find_best_device(layer_name, estimated_memory, layer_idx):
...
+            # Get total number of experts
+            num_experts = getattr(config, "num_local_experts", None)  # Mixtral, PhiMoE, Grok, Llama4
+            if num_experts is None:
+                num_experts = getattr(
+                    config, "num_experts", None
+                )  # Qwen2MoE, Olmo, BailingMoE, GroveMoE, HunYuan, LLaDAMoE
+            if num_experts is None:
+                num_experts = getattr(config, "moe_num_primary_experts", None)  # SmallThinker
+            if num_experts is None:
+                num_experts = getattr(config, "n_routed_experts", None)  # DeepSeek
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: if num_experts is none, better use all the linear number in the block

File: auto_round/utils/device.py [hunk 4]
```diff
@@ -732,6 +731,75 @@ def find_best_device(layer_name, estimated_memory, layer_idx):
...
+
+    Examples:
+        - Non-MoE model: returns 1.0
+        - Mixtral (2/8 experts): returns 0.25
+        - Qwen2MoE (4/60 experts): returns ~0.067
+    """
+    from auto_round.utils.model import is_moe
+
+    for name, module in block.named_modules():
+        if is_moe(module):
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: If not is_moe, continue. This way, all the remaining code will have fewer indentation levels, making it easier to read.

File: auto_round/utils/device.py [hunk 5]
```diff
@@ -732,6 +731,75 @@ def find_best_device(layer_name, estimated_memory, layer_idx):
...
+            config = getattr(block, "config", None)
+            if config is None:
+                break
+
+            # Try to get num_experts_per_tok (active experts count)
+            num_experts_per_tok = getattr(
+                config, "num_experts_per_tok", None
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: this code could be refined to be a function,  get_attr(config, list[str]) and use for

File: auto_round/compressors/utils.py (hunk 1 above)
Comment (wenhuach21) [reply auto_round/compressors/utils.py]: If the parameter is already on the CPU, it won’t be copied but will instead share the same memory.

File: auto_round/utils/device.py (hunk 3 above)
Comment (xin3he) [reply auto_round/utils/device.py]: Yes, if any required value is None, we set moe_ratio=1.0 meaning all linear is considered.

File: auto_round/compressors/base.py [hunk 6]
```diff
@@ -2644,7 +2654,6 @@ def _quantize_block(
             set_amax_for_all_moe_layers(block, attr_name="orig_layer.act_max")
 
         if self.enable_quanted_input:
-            clear_memory()
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: Why remove this clear_memory? All such calls are meant to be added, not omitted.

Comment (xin3he) [reply auto_round/compressors/base.py]: I moved it to a upper place, right after tuning end. Because unwrapper model also requires cuda memory.

Comment (xin3he) [reply auto_round/compressors/base.py]: I add self.low_gpu_memory_usage after moved to upper place, what do you think?

File: auto_round/compressors/base.py [hunk 7]
```diff
@@ -2644,7 +2651,6 @@ def _quantize_block(
...
```
Comment (xin3he) [inline auto_round/compressors/base.py]: fixed

File: auto_round/compressors/base.py [hunk 8]
```diff
@@ -2614,21 +2612,29 @@ def _quantize_block(
...
+                    clear_memory_if_reached_threshold(threshold=0.8)
 
             if i == 0:
                 init_loss = total_loss
 
             if total_loss < best_loss:
                 best_loss = total_loss
                 if not self.not_use_best_mse:
-                    best_params = collect_best_params(block)
+                    best_params = collect_best_params(block, self.low_gpu_mem_usage)
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: better use cache_devcie, which is more clearer.

File: auto_round/compressors/base.py [hunk 9]
```diff
@@ -2413,6 +2413,7 @@ def _get_current_q_output(
         input_others: dict,
         indices: list[int],
         device: str,
+        output_device: str = "cpu",
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: better use cache device

File: auto_round/compressors/base.py [hunk 10]
```diff
@@ -2781,13 +2791,15 @@ def _quantize_blocks(
...
 
+            m.config = model.config if hasattr(model, "config") else None
             q_input, input_ids = quantize_block(
                 m,
                 input_ids,
                 input_others,
                 q_input=q_input,
                 device=device,
             )
+            del m.config
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: if model has no config, this will raise exception

File: auto_round/utils/device.py [hunk 11]
```diff
@@ -412,8 +412,10 @@ def _clear_memory_for_cpu_and_cuda(tensor=None):
         del tensor
     gc.collect()
     if torch.cuda.is_available():
+        torch.cuda.synchronize()
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: rm this if useless

File: auto_round/utils/device.py [hunk 12]
```diff
@@ -821,6 +937,7 @@ def set_auto_device_map_for_block_with_tuning(
     low_gpu_mem_usage=False,
     pick_samples=8,
     output_device=None,
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: type annotation

File: auto_round/wrapper.py [hunk 13]
```diff
@@ -601,7 +601,7 @@ def __init__(self, orig_layer, bit=4, group_size=-1, device="cpu"):
     def unwrapper(self, best_params):
         if best_params is None:
             return self.orig_layer
-        v = best_params["v"]
+        v = best_params["v"].to(self.device)
```
Comment (wenhuach21) [inline auto_round/wrapper.py:604]: have not do the same for the  scale parameters, is it correct?

File: auto_round/compressors/base.py [hunk 14]
```diff
@@ -2458,13 +2459,11 @@ def _quantize_block(
...
-
-        if is_complex_device_mapping(self.device_map):
-            set_auto_device_map_for_block_with_tuning(
-                block, self.device_map, input_ids, self.low_gpu_mem_usage, self.batch_size, device
-            )
-        else:
-            block = block.to(device)
+        # card_0_in_high_risk indicates that card_0 memory is already in high usage (90%) w/o any weights
+        # loss_device is used to calculate loss on the second device if available and card_0_in_high_risk
+        card_0_in_high_risk, loss_device = set_auto_device_map_for_block_with_tuning(
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: The function name doesn’t accurately reflect its behavior. I recommend moving the code that handles the non-auto-device mapping outside the function.

File: auto_round/wrapper.py (hunk 13 above)
Comment (xin3he) [reply auto_round/wrapper.py:604]: In my check, only v is used in the best params
